src/tests/zfs/zfs_ha_controller/zfs_ha_controller_test.go

```diff
@@ -0,0 +1,258 @@
+	Scenario:  Zfs controller in high availability mode
+		Given Zfs controller deployed in non HA mode
+		When Zfs controller is scaled down to zero replica
+		And Zfs volume(pvc) is deployed
+		Then Zfs volume(pvc) should remain in pending state
+		When Zfs controller is scaled to two replica via helm upgrade
+		Then Zfs volume(pvc) should transition to bound state
+		And Deploy fio application to use zfs volume
+		And Taint all worker node with NoSchedule taint
+		When Delete one of the zfs controller pod
+		Then Zfs node lease should point to running zfs controller pod
+*/
+
+var nodeConfig zfs.ZfsNodesDevicePoolConfig
+var defLeaseSwitchTime = 120 // in seconds
+var nodesWithoutTaint []string
+var zfsControllerOrgReplica int32
+var app k8stest.FioApplication
+var volumeProvisionErrorMsg = "waiting for a volume to be created, either by external provisioner"
+
+func controllerHaTest(decor string, engine common.OpenEbsEngine, volType common.VolumeType, fstype common.FileSystemType, volBindModeWait bool) {
+	var err error
+	e2e_config := e2e_config.GetConfig().Product
+
+	app = k8stest.FioApplication{
+		Decor:                          decor,
+		VolSizeMb:                      1024,
+		OpenEbsEngine:                  engine,
+		VolType:                        volType,
+		FsType:                         fstype,
+		Loops:                          5,
+		VolWaitForFirstConsumer:        volBindModeWait,
+		SkipPvcVerificationAfterCreate: true,
+	}
+
+	// setup sc parameters
+	app.Zfs = k8stest.ZfsOptions{
+		PoolName:      nodeConfig.PoolName,
+		ThinProvision: common.No,
+		RecordSize:    "128k",
+		Compression:   common.Off,
+		DedUp:         common.Off,
+	}
+
+	// list nodes without NoSchedule taint
+	nodesWithoutTaint, err = k8stest.ListNodesWithoutNoScheduleTaint()
+	Expect(err).ToNot(HaveOccurred(), "failed to list nodes without NoSchedule taint")
+	logf.Log.Info("nodes without NoSchedule taint", "nodes", nodesWithoutTaint)
+
+	// get the no of replicas in zfs-controller deployment
+	// Scale down the zfs-controller deployment
+	// Check that zfs-controller pods has been terminated successfully
+	logf.Log.Info("Scale down and get the no of replicas in zfs-controller deployment")
+	zfsControllerName := e2e_config.ZfsEngineControllerDeploymentName
+	zfsControllerOrgReplica, err = k8sinstall.ScaleZfsControllerViaHelm(0)
+	Expect(err).To(BeNil(), "failed to scale down deployment %s, error: %v", zfsControllerName, err)
+	Expect(zfsControllerOrgReplica).ShouldNot(BeZero(), "zfs controller replica count should not be zero")
+	logf.Log.Info("zfs controller deployment", "name", zfsControllerName, "original replica", zfsControllerOrgReplica)
+
+	// create sc and pvc
+	logf.Log.Info("create sc, pvc, fio pod with snapshot as source")
+	err = app.CreateVolume()
+	Expect(err).To(BeNil(), "failed to create pvc")
+
+	// verify pvc to be in pending state
+	pvcPhase, err := k8stest.GetPvcStatusPhase(app.GetPvcName(), common.NSDefault)
```

Comment:
Given the we wait for pvc pre condition failed event - is this check required?

Reply:
PVC phase checking might be good because , pvc event we verify is not failed but it's Normal. So, it's better to verify pvc phase before checking normal pvc event .

Reply:
Don't we need to wait for while before asserting the PVC phase is Pending?

Reply:
correct, we can put sleep of 30 seconds before checking pvc state

Reply:
[added 30s sleep ](https://github.com/openebs/openebs-e2e/compare/fdcab88233301473e1f0ba4206b84d5193904e28..1162261ac57cab26e57a485906d2ce672e042dfd)